pkg/k2/constructs/template/inputs/properties_template.go

```diff
@@ -0,0 +1,430 @@
+				return nil, fmt.Errorf("property must be 'map' or 'key_value_list' (was %s) for %s", p.Type, fieldName)
+			}
+		case "ItemProperty":
+			hasItemType := strings.Contains(p.Type, "(")
+			elemProp := srcField.Interface().(*InputTemplate)
+			if hasItemType {
+				itemType := strings.TrimSuffix(
+					strings.TrimPrefix(strings.TrimPrefix(p.Type, "list("), "set("),
+					")",
+				)
+				if elemProp.Type != "" && elemProp.Type != itemType {
+					return nil, fmt.Errorf("item property type must be %s (was %s)", itemType, elemProp.Type)
+				} else if elemProp.Type == "" {
+					elemProp.Type = itemType
+				}
+			}
+			converted, err := elemProp.Convert()
+			if err != nil {
+				return nil, fmt.Errorf("could not convert %s: %w", fieldName, err)
+			}
+			srcField = reflect.ValueOf(converted)
+		}
+
+		if srcField.Type().AssignableTo(dstField.Type()) {
+			dstField.Set(srcField)
+			continue
+		}
+
+		if dstField.Kind() == reflect.Ptr && srcField.Kind() == reflect.Ptr {
+			if srcField.Type().Elem().AssignableTo(dstField.Type().Elem()) {
+				dstField.Set(srcField)
+				continue
+			} else if srcField.Type().Elem().ConvertibleTo(dstField.Type().Elem()) {
+				val := srcField.Elem().Convert(dstField.Type().Elem())
+				// set dest field to a pointer of val
+				dstField.Set(reflect.New(dstField.Type().Elem()))
+				dstField.Elem().Set(val)
+				continue
+			}
+		}
+
+		if conversion, found := fieldConversion[fieldName]; found {
+			err := conversion(srcField, p, propertyType)
+			if err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		return nil, fmt.Errorf(
+			"could not assign %s#%s (%s) to field in %T (%s)",
+			p.Path, fieldName, srcField.Type(), propertyType, dstField.Type(),
+		)
+
+	}
+
+	return propertyType, nil
+}
+
+func setChildPaths(property *InputTemplate, currPath string) {
+	for name, child := range property.Properties {
+		child.Name = name
+		path := currPath + "." + name
+		child.Path = path
+		setChildPaths(child, path)
+	}
+}
+
+func (p InputTemplateMap) Clone() InputTemplateMap {
+	newProps := make(InputTemplateMap, len(p))
+	for k, v := range p {
+		newProps[k] = v.Clone()
+	}
+	return newProps
+}
+
+func (p *InputTemplate) Clone() *InputTemplate {
+	cloned := *p
+	cloned.Properties = make(InputTemplateMap, len(p.Properties))
+	for k, v := range p.Properties {
+		cloned.Properties[k] = v.Clone()
+	}
+	return &cloned
+}
+
+// fieldConversion is a map providing functionality on how to convert inputs into our internal types if they are not inherently the same structure
+var fieldConversion = map[string]func(val reflect.Value, p *InputTemplate, kp property.Property) error{
+	"SanitizeTmpl": func(val reflect.Value, p *InputTemplate, kp property.Property) error {
+		sanitizeTmpl, ok := val.Interface().(string)
+		if !ok {
+			return fmt.Errorf("invalid sanitize template")
+		}
+		if sanitizeTmpl == "" {
+			return nil
+		}
+		// generate random uuid as the name of the template
+		name := uuid.New().String()
```

Comment:
Why use a uuid and not the property's name? Since we don't have delegate templates, the only place this is relevant is in making the error messages.